### ISSUE TITLE
Training new AuxiliaryASR and PitchExtractor

### ISSUE THREAD

**Issue by shaun95** [yl4579/StyleTTS#9]
Great work, thanks.

I wish to train the [PitchExtractor](https://github.com/yl4579/PitchExtractor) and [AuxiliaryASR ](https://github.com/yl4579/AuxiliaryASR) from scratch and load the new models with StyleTTS using [TMA_CEloss](https://github.com/yl4579/StyleTTS/blob/main/Configs/config.yml#L49) = True

I see the model.py graph for each of the two separate repos is different from the code provided under StyleTTS/Utils/, and that made me think if I train each of PitchExtractor and AuxiliaryASR they won't load in StyleTTS, that is unless I make them identical to the models from StyleTTS.  Is that correct?

**Comment by yl4579** [comment]
For your first question, you will need to change the SOS index in [model.py](https://github.com/yl4579/AuxiliaryASR/blob/5cb3ee46c4839d96d50c8ad9b3f378b1d7f06df7/models.py), but as you said, it doesn't matter if the semicolon never appears in your text.

For your second question, token 0 in StyleTTS was used as SOS and EOS instead of blank. Blank is index 16. You can see there is no SOS and EOS in StyleTTS. Also, for the text aligner, we removed the SOS token at the beginning of the attention, so it doesn't matter in the end.

**Comment by shaun95** [comment]
Thanks.

**Comment by lexkoro** [comment]
@shaun95 Were you able to get reasonable results by training everything from scratch? 
I retrained AuxiliaryASR and PitchExtractor, but training StyleTTS I am not able to achieve good results for a multi-speaker dataset I am using.

**Comment by shaun95** [comment]
@lexkoro I was able to train the necessary models from scratch, I had to make some guesses on some code on the AuxiliaryASR, at the end while the StyleTTS speaks, there are some quality issues like short pauses inserted between words, and some phonemes repeat. I am guessing it has to do with the TMA code I trained from scratch with my adaptations. It's a pity as it takes a long time to get all of these separate models to train and the quality I get is not as the reference implementation. Incidentally, I am using different speaker dataset and I also am not trying for multiple speakers. Just trying a single speaker.

**Comment by yl4579** [comment]
@lexkoro Do you mean your ASR and F0 model performs worse compared to those released by me, on the same dataset? 

@shaun95 The training code should be the same for the ASR models regardless of how StyleTTS uses it. I only modified the inference code to incorporate the CE loss for alignment, which was unimportant in practice. I have the recipe for official Hifi-GAN and BIGVGAN in 22.05 kHz if you need that. It works as well as the demo on LJSpeech, though I don't have time to test on multispeaker as I'm working on E2E training now. It will be named StyleTTS 2, where you won't need to retrain any components for E2E training, though you will need more RAM for that. The quality is also much better with E2E training, so you stay tuned for that.

**Comment by shaun95** [comment]
@yl4579 @yl4579, thank you for your response. The audio samples you provided for StyleTTS are excellent My goal was to replicate the training steps in StyleTTS so that I could test a new voice on a model trained on a single speaker, without relying on the provided checkpoints for Pitch or ASR. To achieve this, I trained a new AuxiliaryASR and PitchExtractor using the original training repositories as indicated in the StyleTTS readme file. However, I faced some issues after the training of each was completed and tried to load them into StyleTTS, so I decided to make some changes to the code in each training repository. For instance, I assumed that the dropouts applied in the Pitch training repository should be the same as the ones done in StyleTTS/Utils/JDC. If possible, it would be helpful and much appreciated if the respective training scripts for Pitch and ASR were provided in the same repository when working on StyleTTS2 for simplicity's sake. Once again, thank you for this repository, I appreciate it greatly. I look forward to StyleTTS2

**Comment by lexkoro** [comment]
@shaun95 Thanks for the reply! I experience the same and I know from at least another person who was also not able to reproduce the quality as provided in the examples. 

@yl4579 Hey, no the newly trained ASR and F0 seemed fine looking at the losses. Training ASR was a bit challenging, since it would sometimes NaN after the first few steps. I trained on a highly expressive dataset from video games in the languages german, english, polish, russian and japanese.

Now I tried training StyleTTS on a small subset of the data used for training F0 and ASR. But I am not able to achieve satisfying results. As @shaun95 mentioned the style transfer works quite well but the quality doesn't seem to improve or rather said is nowhere near as the released examples.

Training E2E will be a welcome change, since training all those components is quite cumbersome! :)
Thanks for open-sourcing your code! Even if I won't be able to reproduce the results with my own data, I really appreciate you sharing the project! 👍

**Comment by yl4579** [comment]
@shaun95 You don't need to modify anything. You can just train it with the exact same code I pushed and load that checkpoint for StyleTTS. There's no dropout in StyleTTS because during inference we do not need those things. They were only used for training. 

@lexkoro That is weird because all the models I released we trained with exactly the same code I pushed to the repo. I also trained the ASR and F0 models with HifiGAN recipes in 22.05 kHz and was able to reproduce the results on LJSpeech, so I believe the code should be good. I'm wondering if it could be a problem with the dataset. Can you maybe test it with the pre-trained model directly on your dataset, without changing anything?

**Comment by shaun95** [comment]
@yl4579 OK, I am going to re-examine my steps. What about the use of g2p_enu versus phonemizer in text-aligner (StyleTTS vs ARR)? Don't I need to account for that or it does not matter. I appreciate you feedback!

**Comment by lexkoro** [comment]
@yl4579 Testing with the pre-trained model the synthesized quality is much better. Of course the speaker similarity is sometimes quite of which makes sense since it's zero shot in this case. 

So now I am wondering where the problem in my training might stem from. I have modified the meldataset.py in each module to use the mel-processing from BigVGAN.
Additionally I tried increasing the style_dim to 256 and dim_in to 128, since I have a high variation in speech, not sure this one even makes sense. But the problem also persisted with the default settings.

Personally it feels like that in my case the pitch of the synthesized speech is somewhat off making the voices either to high or to low pitched and it sounds wobbly/ scratchy.

How long did you train the LibriTTS model? Judging by the config it did converge quite fast.

My changes can be found here https://github.com/lexkoro/StyleTTS

**Comment by shaun95** [comment]
@lexkoro, based on my personal experience, I didn't encounter any issues with the intonation being too high or low during my experiments. The short utterances sounded good, but I noticed a decline in quality for medium to long utterances. Specifically, I observed unnatural pacing of words with pauses between them, and sometimes mumbling occurred. However, the emotion style transfer worked well overall. 

During my training, I trained all models from scratch three times consecutively, once with TMA_CE loss set on and twice with it off. Although I didn't achieve satisfactory alignment during training with StyleTTS when using TMA_CE set to True, it was still better than when not using it. Additionally, I trained a new HiFiGAN model on the speaker dataset, and the synthesizer worked well without any signal issues. When judging the vocoder alone, the speech output sounded fine. 

However, I do think the pacing of the words and mumbling could indicate an issue with the alignment, which may be related to my training of AuxiliaryASR or StyleTTS.

**Comment by yl4579** [comment]
@shaun95 Can you look at the alignment from the Tensorboard log to see if it looks monotonic?

**Comment by shaun95** [comment]
@yl4579 The alignment was monotonic in the two instances when I set TMA_CE to False. The output of these trainings is what I use for my answers here. Both of these attempts sounded similar.  When TMA_CE was set to True I didn't get an alignment and the speech output was garbled. Thanks for your feedback

**Comment by yl4579** [comment]
@shaun95 Sorry for the late reply. I didn't see this until now. Hope your problem was resolved, but if not, here is my two cents for your problem. How was the quality of the model after the first stage, like did the reconstruction work well?

@RasmusD The change was to make the alignment over the phoneme axis instead of over the mel axis. In the AuxiliaryASR, it was trained to align melspectrograms with texts (i.e., the input is melspectrograms and the output is the text) because it is an ASR model, while in TTS we want the input to be texts and the output is the melspectrogram. The latter is the reverse of the problem, but you cannot simply transpose the attention matrix because the attention for AuxiliaryASR was normalized across the melspectrogram frames instead of the phoneme tokens. I changed it specifically to renormalize it in the correct axis. 

As for the pitch extractor, it was my neglect that I didn't add an inference method but just changing the model directly, it was just for simplicity, as there should be no dropout during inference. The shape of input during training of the F0 model was fixed to be 192, while the shape of input for TTS should be variable, so I removed that constraint.

**Comment by RasmusD** [comment]
That's great - thanks a lot for the clarifications!

**Comment by glimperg** [comment]
I'm currently trying to replicate the results from your pretrained model, but I'm not getting the expected results when training StyleTTS using an AuxiliaryASR model that I trained myself. When using the pretrained ASR model provided in the repo, I get diagonal alignments after training StyleTTS for only one epoch:
![pretrained_asr_styletts_onepoch](https://github.com/yl4579/StyleTTS/assets/25743672/affcd889-d642-421c-bef1-004b6cc3b16c)
 However, when training AuxiliaryASR myself, the alignments are incorrect after even 20 epochs:
![selftrained_asr_styletts_20epochs](https://github.com/yl4579/StyleTTS/assets/25743672/2eeb4fe4-baf7-4c9d-9e82-c947fb4e2613)
For training the AuxiliaryASR model, I used the config from this repo provided in `Utils/ASR/config.yml`, and I trained it on the files from `train_list.txt` in the AuxiliaryASR repo (i.e. the subset of LJSpeech). The strange thing is that it seems that the ASR model trained correctly, looking at the eval alignments logged to tensorboard while training:
![auxiliaryasr_80epochs](https://github.com/yl4579/StyleTTS/assets/25743672/987024bb-49ba-4cf2-98cd-d7b6658b450e)

Any ideas @yl4579 what might have gone wrong here? Should I have trained AuxiliaryASR on a different dataset, did I not use the correct config file for AuxiliaryASR, or is it something else?

In any case, thanks for providing the code and models for this project!
